src/sema/eval.rs

```diff
@@ -196,3 +202,398 @@ pub fn eval_const_rational(
         )),
     }
 }
+
+/// Helper function that recurses the expression, and calls expression = eval_constants_in_expression().
+/// If expression is an arithmetic operation of two number literals, it will be stored in an expression list to later be evaluated by verify_expression_for_overflow().
+/// If the return is a diagnostic, store it in a diagnostics list.
+fn check_term_for_constant_overflow(expr: &Expression, params: &mut RecurseParams) -> bool {
+    match expr {
+        Expression::Add(..)
+        | Expression::Subtract(..)
+        | Expression::Multiply(..)
+        | Expression::Divide(..)
+        | Expression::Modulo(..)
+        | Expression::Power(..)
+        | Expression::ShiftLeft(..)
+        | Expression::ShiftRight(..)
+        | Expression::BitwiseAnd(..)
+        | Expression::BitwiseOr(..)
+        | Expression::BitwiseXor(..) => match eval_constants_in_expression(expr) {
+            Ok(expression) => {
+                if let Expression::NumberLiteral(..) = expression {
+                    params.recursed.push(expression);
+                    return false;
+                }
+            }
+            Err(diagnostic) => {
+                params.diagnostics.push(diagnostic);
+            }
+        },
+        _ => {
+            if let Expression::NumberLiteral(..) = expr {
+                params.recursed.push(expr.clone());
+                return false;
+            }
+        }
+    }
+
+    true
+}
+
+/// This function recursively folds number literals in a given expression.
+fn eval_constants_in_expression(expr: &Expression) -> Result<Expression, Diagnostic> {
+    match expr {
+        Expression::Add(loc, ty, unchecked, left, right) => {
+            let left = eval_constants_in_expression(&*left)?;
+            let right = eval_constants_in_expression(&*right)?;
+
+            if let (Expression::NumberLiteral(_, _, left), Expression::NumberLiteral(_, _, right)) =
+                (&left, &right)
+            {
+                Ok(Expression::NumberLiteral(*loc, ty.clone(), left.add(right)))
+            } else {
+                Ok(Expression::Add(
+                    *loc,
+                    ty.clone(),
+                    *unchecked,
+                    Box::new(left),
+                    Box::new(right),
+                ))
```

Comment:
This is a non-constant expression, which is not used anywhere I think. 

How about returning making the return value `Result<Option<Expression>, Diagnostic>` and then returning `Ok(None)`